### PR TITLE
Korean translation for chapter 2

### PR DESCRIPTION
**Pull request by JOE1994**
Locally checked rendering using `zola` :smiley_cat:

### CODE REVIEW

**Comment by JOE1994** [comment]
CI indicates there are some broken URLs (from both the original post & the translation posts)

**Comment by JOE1994** [comment]
I will try to find a reviewer for this translation by posting a message to the Discord chat of [Korean Rust users' group](https://rust-kr.org/).

**Comment by phil-opp** [comment]
Thanks a lot for the translation @JOE1994! Any luck with finding someone to review it?

@QuqqU Maybe you could take a quick look?

**Comment by JOE1994** [comment]
> Thanks a lot for the translation @JOE1994! Any luck with finding someone to review it?

Unfortunately not yet 😢 
It seems like people are busy with final exam preps, other  ongoing projects, etc..

**Comment by phil-opp** [comment]
Ok, no problem. If we don't find anyone within the next week, I will add it to "call for participation" section of the next "This Month in Rust OSDev" issue.

**Comment by QuqqU** [comment]
I think I can do it until this weekend!
I couldn't review it quickly because I had a lot of work to do this month, but it's almost over.

File: blog/content/edition-2/posts/02-minimal-rust-kernel/disable-red-zone/index.ko.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+[red zone]: https://eli.thegreenplace.net/2011/09/06/stack-frame-layout-on-x86-64#the-red-zone
+[System V ABI]: https://wiki.osdev.org/System_V_ABI
+
+<!-- more -->
+
+![stack frame with red zone](red-zone.svg)
+
+위 사진은 `n`개의 지역 변수를 가진 함수의 스택 프레임을 보여줍니다. 함수가 호출되었을 때, 스택에 함수의 반환 주소 및 지역 변수들을 스택에 저장할 수 있도록 스택 포인터의 값이 조정됩니다.
+
+red zone은 조정된 스택 포인터 아래의 128바이트의 메모리 구간을 가리킵니다. 함수가 또다른 함수를 호출하지 않는 구간에서만 사용하는 임시 데이터의 경우, 함수가 이 구간에 해당 데이터를 저장하는 데 이용할 수 있습니다. 따라서 스택 포인터를 조정하기 위해 필요한 명령어 두 개를 생략할 수 있는 상황이 종종 있습니다 (예: 다른 함수를 호출하지 않는 작은 함수).
```
Comment (QuqqU) [inline blog/content/edition-2/posts/02-minimal-rust-kernel/disable-red-zone/index.ko.md:19]: 또다른 -> 또 다른

Comment (QuqqU) [inline blog/content/edition-2/posts/02-minimal-rust-kernel/disable-red-zone/index.ko.md:19]: 또다른 -> 또 다른

File: blog/content/edition-2/posts/02-minimal-rust-kernel/disable-red-zone/index.ko.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+하지만 이 최적화 기법을 사용하는 도중 소프트웨어 예외(exception) 혹은 하드웨어 인터럽트가 일어날 경우 큰 문제가 생깁니다. 함수가 red zone을 사용하던 도중 예외가 발생한 상황을 가정해보겠습니다:
+
+![red zone overwritten by exception handler](red-zone-overwrite.svg)
+
+CPU와 예외 처리 핸들러가 red zone에 있는 데이터를 덮어씁니다. 하지만 이 데이터는 인터럽트된 함수가 사용 중이었던 것입니다. 따라서 예외 처리 핸들러로부터 반환하여 다시 인터럽트된 함수가 계속 실행되게 되었을 때 변경된 red zone의 데이터로 인해 함수가 오작동할 수 있습니다. 이런 현상으로 인해 [디버깅 하는 데에 몇 주씩 걸릴 수 있는 이상한 버그][take weeks to debug]가 발생할지도 모릅니다.
```
Comment (QuqqU) [inline blog/content/edition-2/posts/02-minimal-rust-kernel/disable-red-zone/index.ko.md:25]: 디버깅 하는 -> 디버깅하는 or 디버깅을 하는

File: blog/content/edition-2/posts/02-minimal-rust-kernel/disable-red-zone/index.ko.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+[take weeks to debug]: https://forum.osdev.org/viewtopic.php?t=21720
+
+미래에 예외 처리 로직을 구현할 때 이러한 오류가 일어나는 것을 피하기 위해 우리는 미리 red zone 최적화 기법을 해제한 채로 프로젝트를 진행할 것입니다. 컴파일 대상 환경 설정 파일에 `"disable-redzone": true` 라는 줄을 추가함으로써 해당 기능을 해제할 수 있습니다.
```
Comment (QuqqU) [inline blog/content/edition-2/posts/02-minimal-rust-kernel/disable-red-zone/index.ko.md:29]: `"disable-redzone": true` 라는 줄을 -> `"disable-redzone": true` 줄을

위와 같이 간결하게 쓰는 것이 독자가 자연스럽게 읽는 것 같습니다.

File: blog/content/edition-2/posts/02-minimal-rust-kernel/disable-red-zone/index.ko.md (hunk 1 above)
Comment (QuqqU) [inline blog/content/edition-2/posts/02-minimal-rust-kernel/disable-red-zone/index.ko.md:19]: small leaf functions 를 직역하면 '다른 함수를 호출하지 않는 작은 함수'가 되지만 '작은'이라는 키워드가 이해를 어렵게 하는 것 같습니다.
'다른 함수를 호출하지 않는 작은 함수'를 '다른 함수를 호출하지 않는 함수'로 일부 생략해 표현하는 의견을 제시합니다

File: blog/content/edition-2/posts/02-minimal-rust-kernel/disable-simd/index.ko.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
++++
+title = "SIMD 해제하기"
+weight = 2
+path = "ko/disable-simd"
+template = "edition-2/extra.html"
++++
+
+[Single Instruction Multiple Data (SIMD)] 명령어들은 여러 데이터 word에 동시에 덧셈 등의 작업을 실행할 수 있으며, 이를 통해 프로그램의 실행 시간을 상당히 단축시킬 수 있습니다. `x86_64` 아키텍처는 다양한 SIMD 표준들을 지원합니다:
```
Comment (QuqqU) [inline blog/content/edition-2/posts/02-minimal-rust-kernel/disable-simd/index.ko.md:8]: 단축시킬 -> 단축할

File: blog/content/edition-2/posts/02-minimal-rust-kernel/disable-simd/index.ko.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+- [MMX]: _Multi Media Extension_ 명령어 집합은 1997년에 등장하였으며, `mm0`에서 `mm7`까지 8개의 64비트 레지스터들을 정의합니다. 이 레지스터들은 그저 [x87 부동 소수점 장치]의 레지스터들을 가리키는 별칭입니다.
+- [SSE]: _Streaming SIMD Extensions_ 명령어 집합은 1999년에 등장하였습니다. 부동 소수점 연산용 레지스터를 재사용하는 대신 새로운 레지스터 집합을 도입했습니다. `xmm0`에서 `xmm15`까지 16개의 새로운 128비트 레지스터를 정의합니다.
+- [AVX]: _Advanced Vector Extensions_ 은 SSE에 추가로 멀티미디어 레지스터의 크기를 늘리는 확장 표준입니다. `ymm0`에서 `ymm15`까지 16개의 새로운 256비트 레지스터를 정의합니다. `ymm` 레지스터들은 기존의 `xmm` 레지스터를 확장합니다 (`xmm0`이 `ymm0` 레지스터의 하부 절반을 차지하는 식으로 다른 15개의 짝에도 같은 방식의 확장이 적용됩니다).
+
+[MMX]: https://en.wikipedia.org/wiki/MMX_(instruction_set)
+[x87 floating point unit]: https://en.wikipedia.org/wiki/X87
+[SSE]: https://en.wikipedia.org/wiki/Streaming_SIMD_Extensions
+[AVX]: https://en.wikipedia.org/wiki/Advanced_Vector_Extensions
+
+이러한 SIMD 표준들을 사용하면 프로그램 실행 속도를 많이 향상시킬 수 있는 경우가 많습니다. 우수한 컴파일러는 [자동 벡터화 (auto-vectorization)][auto-vectorization]이라는 과정을 통해 일반적인 반복문을 SIMD 코드로 변환할 수 있습니다.
```
Comment (QuqqU) [inline blog/content/edition-2/posts/02-minimal-rust-kernel/disable-simd/index.ko.md:23]: 향상시킬 수 -> 향상할 수

File: blog/content/edition-2/posts/02-minimal-rust-kernel/disable-simd/index.ko.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+컴파일 대상 환경 설정 파일의 `features` 필드를 이용해 해당 기능들을 해제할 수 있습니다. `mmx` 및 `sse` 기능을 해제하려면 아래와 같이 해당 기능 이름 앞에 빼기 기호를 붙여주면 됩니다:
+
+```json
+"features": "-mmx,-sse"
+```
+
+## 부동소수점 (Floating Point)
+
+우리의 입장에서는 안타깝게도, `x86_64` 아키텍처는 부동 소수점 계산에 SSE 레지스터를 사용합니다. 따라서 SSE 기능이 해제된 상태에서 부동 소수점 계산을 컴파일 하려고 하면 LLVM이 오류를 일으킵니다. Rust의 core 라이브러리는 이미 부동 소수점 숫자들을 사용하기에 (예: `f32` 및 `f64` 에 대한 각종 trait들을 정의함), 우리의 커널에서 부동 소수점 계산을 피하더라도 부동 소수점 계산을 컴파일하는 것을 피할 수 없습니다.
```
Comment (QuqqU) [inline blog/content/edition-2/posts/02-minimal-rust-kernel/disable-simd/index.ko.md:37]: 컴파일 하려고 하면 -> 컴파일하면

File: blog/content/edition-2/posts/02-minimal-rust-kernel/index.ko.md [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+### BIOS 부팅
+
+UEFI 표준으로 동작하는 최신 기기들도 가상 BIOS를 지원하기에, 존재하는 거의 모든 x86 시스템들이 BIOS 부팅을 지원합니다. 덕분에 하나의 BIOS 부팅 로직을 구현하면 여태 만들어진 거의 모든 컴퓨터를 부팅시킬 수 있습니다. 동시에 이 방대한 호환성이 BIOS의 가장 큰 약점이기도 한데,
+그 이유는 1980년대의 구식 부트로더들에 대한 하위 호환성을 유지하기 위해 부팅 전에는 항상 CPU를 16비트 호환 모드 ([real mode]라고도 불림)로 설정해야 하기 때문입니다.
+
+이제 BIOS 부팅 과정의 첫 단계부터 살펴보겠습니다:
+
+여러분이 컴퓨터의 전원을 켜면, 제일 먼저 컴퓨터는 마더보드의 특별한 플래시 메모리로부터 BIOS 이미지를 로드합니다. BIOS 이미지는 자가 점검 및 하드웨어 초기화 작업을 처리한 후에 부팅 가능한 디스크가 있는지 탐색합니다. 부팅 가능한 디스크가 있다면, 제어 흐름은 해당 디스크의 _부트로더 (bootloader)_ 에게 넘겨집니다. 이 부트로더는 디스크의 가장 앞 주소 영역에 저장되는 512 바이트 크기의 실행 파일입니다. 대부분의 부트로더들의 경우 로직을 저장하는 데에 512 바이트보다 더 큰 용량이 필요하기에, 부트로더의 로직을 둘로 쪼개어 첫 단계 로직을 첫 512 바이트 안에 담고, 두 번째 단계 로직은 첫 단계 로직에 의해 로드된 이후 실행됩니다.
+
+부트로더는 커널 이미지가 디스크의 어느 주소에 저장되어있는지 알아낸 후 메모리에 커널 이미지를 로드해야 합니다. 그 다음 CPU를 16비트 [real mode]에서 32비트 [protected mode]로 전환시키고, 그 후에 다시 CPU를 64비트 [long mode]로 전환시킨 이후부터 64비트 레지스터 및 메인 메모리의 모든 주소를 사용할 수 있게 됩니다. 부트로더가 세 번째로 할 일은 BIOS로부터 메모리 매핑 정보 등의 필요한 정보를 알아내어 운영체제 커널에 전달하는 것입니다.
```
Comment (QuqqU) [inline blog/content/edition-2/posts/02-minimal-rust-kernel/index.ko.md:55]: 그 다음 -> 그다음
전환시키고 -> 전환하고
전환시킨 이후부터 -> 전환한 이후부터

File: blog/content/edition-2/posts/02-minimal-rust-kernel/index.ko.md [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+
+[real mode]: https://en.wikipedia.org/wiki/Real_mode
+[protected mode]: https://en.wikipedia.org/wiki/Protected_mode
+[long mode]: https://en.wikipedia.org/wiki/Long_mode
+[memory segmentation]: https://en.wikipedia.org/wiki/X86_memory_segmentation
+
+부트로더를 작성하는 것은 상당히 성가신 작업인데, 그 이유는 어셈블리 코드도 작성해야 하고 "A 레지스터에 B 값을 저장하세요" 와 같이 원리를 단번에 이해하기 힘든 작업들이 많이 수반되기 때문입니다. 따라서 이 포스트에서는 부트로더를 만드는 것 자체를 다루지는 않고, 대신 운영체제 커널의 맨 앞에 부트로더를 자동으로 추가해주는 [bootimage]라는 도구를 제공합니다.
```
Comment (QuqqU) [inline blog/content/edition-2/posts/02-minimal-rust-kernel/index.ko.md:62]: 작업들이 -> 작업이

File: blog/content/edition-2/posts/02-minimal-rust-kernel/index.ko.md [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+
+[disabling the red zone]: @/edition-2/posts/02-minimal-rust-kernel/disable-red-zone/index.ko.md
+
+```json
+"features": "-mmx,-sse,+soft-float",
+```
+
+`features` 필드는 컴파일 대상 환경의 기능들을 활성화/비활성화 하는 데 이용합니다. 우리는 `-` 기호를 통해 `mmx`와 `sse` 기능들을 비활성화시키고 `+` 기호를 통해 `soft-float` 기능을 활성화시킬 것입니다. `features` 필드의 문자열 내부 플래그들 사이에 빈칸이 없도록 해야 합니다. 그렇지 않으면 LLVM이 `features` 필드의 문자열 값을 제대로 해석하지 못하기 때문입니다.
+
+`mmx`와 `sse`는 [Single Instruction Multiple Data (SIMD)] 명령어들의 사용 여부를 결정하는데, 해당 명령어들은 프로그램의 실행 속도를 훨씬 빠르게 만드는 데에 도움을 줄 수 있습니다. 하지만 운영체제에서 큰 SIMD 레지스터를 사용할 경우 커널의 성능에 문제가 생길 수 있습니다. 그 이유는 커널이 인터럽트 되었던 프로그램을 다시 실행시키기 전에 모든 레지스터 값들을 인터럽트 직전 시점의 상태로 복원시켜야 하기 때문입니다. 커널이 SIMD 레지스터를 사용하려면 각 시스템 콜 및 하드웨어 인터럽트가 일어날 때마다 모든 SIMD 레지스터에 저장된 값들을 메인 메모리에 저장해야 할 것입니다. SIMD 레지스터들이 총 차지하는 용량은 매우 크고 (512-1600 바이트) 인터럽트 또한 자주 일어날 수 있기에,
```
Comment (QuqqU) [inline blog/content/edition-2/posts/02-minimal-rust-kernel/index.ko.md:197]: 다시 실행시키기 전에 -> 다시 실행하기 전에

File: blog/content/edition-2/posts/02-minimal-rust-kernel/index.ko.md [hunk 10]
```diff
@@ -0,0 +1,511 @@
...
+   Compiling compiler_builtins v0.1.32
+   Compiling blog_os v0.1.0 (/…/blog_os)
+    Finished dev [unoptimized + debuginfo] target(s) in 0.29 secs
+```
+
+이제 `cargo build` 명령어가 `core`, `rustc-std-workspace-cord` (`compiler_builtins`가 필요로 하는 라이브러리) 그리고 `compiler_builtins` 라이브러리를 우리의 커스텀 컴파일 대상을 위해 다시 컴파일하는 것을 확인할 수 있습니다.
+
+#### 메모리 관련 내장 함수
+
+Rust 컴파일러는 특정 군의 내장 함수들이 (built-in function) 모든 시스템에서 주어진다고 가정합니다. 대부분의 내장 함수들은 우리가 방금 컴파일한 `compiler_builtins` 크레이트가 이미 갖추고 있습니다. 하지만 그 중 몇몇 메모리 관련 함수들은 기본적으로 사용 해제 상태가 되어 있는데, 그 이유는 해당 함수들을 호스트 시스템의 C 라이브러리가 제공하는 것이 관례이기 때문입니다. `memset`(메모리 블럭 전체에 특정 값 저장하기), `memcpy` (한 메모리 블럭의 데이터를 다른 메모리 블럭에 옮겨쓰기), `memcmp` (메모리 블럭 두 개의 데이터를 비교하기) 등이 이 분류에 해당됩니다. 여태까지는 우리가 이 함수들 중 어느 하나도 사용하지 않았지만, 운영체제 구현을 더 추가하다 보면 필수적으로 사용될 함수들입니다 (예를 들어, 구조체를 복사하여 다른 곳에 저장할 때).
```
Comment (QuqqU) [inline blog/content/edition-2/posts/02-minimal-rust-kernel/index.ko.md:309]: 해당됩니다 -> 해당합니다
그 중 -> 그중

File: blog/content/edition-2/posts/02-minimal-rust-kernel/index.ko.md [hunk 11]
```diff
@@ -0,0 +1,511 @@
...
+이제 이 함수에 코드를 추가하여 화면에 메세지를 출력해볼 것입니다.
+
+### 화면에 출력하기
+현재 단계에서 가장 쉽게 화면에 문자를 출력할 수 있는 방법은 바로 [VGA 텍스트 버퍼][VGA text buffer]를 이용하는 것입니다. 이것은 VGA 하드웨어에 매핑되는 특수한 메모리 영역이며 화면에 출력될 내용이 저장됩니다. 주로 이 버퍼는 주로 25행 80열 (행마다 80개의 문자 저장)로 구성됩니다. 각 문자는 ASCII 문자로서 전경색 혹은 배경색과 함께 화면에 출력됩니다. 화면 출력 결과의 모습은 아래와 같습니다:
+
+[VGA text buffer]: https://en.wikipedia.org/wiki/VGA-compatible_text_mode
+
+![ASCII 문자들을 출력한 화면의 모습](https://upload.wikimedia.org/wikipedia/commons/f/f8/Codepage-437.png)
+
+VGA 버퍼가 정확히 어떤 구조를 하고 있는지는 다음 포스트에서 VGA 버퍼 드라이버를 작성하면서 다룰 것입니다. "Hello World!" 메세지를 출력하는 데에는 그저 버퍼의 시작 주소가 `0xb8000`이라는 것, 그리고 각 문자는 ASCII 문자를 위한 1바이트와 색상 표기를 위한 1바이트가 필요하다는 것만 알면 충분합니다.
```
Comment (QuqqU) [inline blog/content/edition-2/posts/02-minimal-rust-kernel/index.ko.md:361]: 메세지 -> 메시지

File: blog/content/edition-2/posts/02-minimal-rust-kernel/index.ko.md [hunk 12]
```diff
@@ -0,0 +1,511 @@
...
+
+USB 드라이브에 우리의 커널을 저장한 후 실제 컴퓨터에서 부팅하는 것도 가능합니다:
+
+```
+> dd if=target/x86_64-blog_os/debug/bootimage-blog_os.bin of=/dev/sdX && sync
+```
+
+`sdX` 대신 여러분이 소지한 USB 드라이브의 기기명을 입력하시면 됩니다. 해당 기기에 쓰인 데이터는 전부 덮어씌워지기 때문에 정확한 기기명을 입력하도록 주의해주세요.
+
+이미지를 USB 드라이브에 다 덮어 썼다면, 이제 실제 하드웨어에서 해당 이미지를 통해 부트하여 실행할 수 있습니다. 아마 특별한 부팅 메뉴를 사용하거나 BIOS 설정에서 부팅 순서를 변경하여 USB로부터 부팅하도록 설정해야 할 것입니다. `bootloader` 크레이트가 아직 UEFI를 지원하지 않기에, UEFI 표준을 사용하는 기기에서는 부팅할 수 없습니다.
```
Comment (QuqqU) [inline blog/content/edition-2/posts/02-minimal-rust-kernel/index.ko.md:486]: 덮어 썼다면 -> 덮어썼다면

File: blog/content/edition-2/posts/02-minimal-rust-kernel/index.ko.md [hunk 13]
```diff
@@ -0,0 +1,511 @@
...
+
+### `cargo run` 명령어 사용하기
+
+QEMU에서 커널을 쉽게 실행할 수 있게 아래처럼 `runner` 라는 새로운 cargo 설정 키 값을 추가합니다.
```
Comment (QuqqU) [inline blog/content/edition-2/posts/02-minimal-rust-kernel/index.ko.md:490]: 'runner' 라는 -> 'runner'라는

**Review by QuqqU — COMMENTED** [review]
I think reviewing using Korean is better to convey meaning of feedback, So leave it in Korean.
As you know, simple comments are for some spacing issue. And (relatively) long comments are for smooth(and maybe better) translation.
I hope you consider it! It was very good time to read your translation :)

**Comment by JOE1994** [comment]
@QuqqU Thank you so much for your time and valuable feedback! Your feedback was very helpful :+1:

**Review by QuqqU — APPROVED** [review]

**Comment by JOE1994** [comment]
I squashed the commits into a single commit, and I think this PR is now ready to be merged.

**Comment by phil-opp** [comment]
Awesome, thanks a lot!

**Comment by phil-opp** [comment]
Published at <https://os.phil-opp.com/ko/minimal-rust-kernel/>

**Pull request merged**